### PR TITLE
c++: Add guid_t struct, implicit conversion ops

### PR DESCRIPTION
**Pull request by r-rojo**
Add a guid_t struct to wrap the guid type.
This uses std::array to store the guid data put also keeps a reference
to a props_ object. It uses the props_ variable to call C API functions
to get/set the GUID property.
It also copies the data when getting and setting so that it doesn't get
stale.

Add unit tests for properties and integrate those in CMake so that one
may use 'make test'

### CODE REVIEW

File: libopae++/include/opaec++/pvalue.h [hunk 1]
```diff
@@ -36,6 +38,57 @@ namespace fpga
 namespace types
 {
 
+struct guid_t
+{
+    guid_t(fpga_properties * p)
```
Comment (tswhison) [inline libopae++/include/opaec++/pvalue.h]: Why fpga_properties * ?  fpga_properties is the opaque type which is already a pointer.

File: libopae++/include/opaec++/pvalue.h [hunk 2]
```diff
@@ -36,6 +38,57 @@ namespace fpga
...
+
+    friend std::ostream & operator<<(std::ostream & ostr, const guid_t & g){
+        char guid_str[84];
+        uuid_unparse(g.data_.data(), guid_str);
+        ostr << guid_str;
+        return ostr;
+    }
+
+private:
+    fpga_properties *props_;
```
Comment (tswhison) [inline libopae++/include/opaec++/pvalue.h]: Same comment as above.

File: libopae++/include/opaec++/pvalue.h [hunk 3]
```diff
@@ -55,44 +108,46 @@ struct pvalue
     {
     }
 
-    template<typename U>
-    pvalue<T>& operator=(U v){
-        T value = static_cast<T>(v);
-        auto res = set_(*props_, value);
-        if (res != FPGA_OK){
-            // TODO: Print error: std::cerr << "Could not set property";
+    template<typename V = typename std::enable_if<!std::is_same<T, guid_t>::value, T>::type>
```
Comment (tswhison) [inline libopae++/include/opaec++/pvalue.h]: Help me understand what this is doing.

File: libopae++/include/opaec++/pvalue.h [hunk 4]
```diff
@@ -55,44 +108,46 @@ struct pvalue
...
-        }else{
-            ostr << "null";
-        }
+        T value = p;
+        ostr << +(value);
         return ostr;
     }
 
 private:
     fpga_properties *props_;
```
Comment (tswhison) [inline libopae++/include/opaec++/pvalue.h]: Why fpga_properties * ?

File: libopae++/include/opaec++/pvalue.h (hunk 1 above)
Comment (r-rojo) [reply libopae++/include/opaec++/pvalue.h]: These property wrapper structs are constructed as members of the `properties` class. They are all initialized with a pointer to the private member `props_` of the `properties` instance which is invalid until the ctor of `properties` which uses `fpgaGetProperties` to allocate the `props_`. That is why all of the property structs need to have a reference (pointer) to `props_`.

File: libopae++/include/opaec++/pvalue.h (hunk 3 above)
Comment (r-rojo) [reply libopae++/include/opaec++/pvalue.h]: Basically, it is enabling this templated function only when `T` is not guid_t. It also assigns the type, `T`, to `V`.
It was something I was experimenting with before when I had `guid_t` a typedef of `fpga_guid`. It's probably not needed anymore and I will most likely remove it an upcoming PR.
`::value` refers to the value of the enable_if (true or false). `::type` refers to the type argument (in this case `T`) of enable_if.

File: libopae++/include/opaec++/pvalue.h (hunk 1 above)
Comment (tswhison) [reply libopae++/include/opaec++/pvalue.h]: Ok, that makes sense.

**Pull request merged**